util/pingpong.c

```
@@ -353,126 +359,161 @@ int pp_getaddrinfo(char *name, uint16_t port, struct addrinfo **results)
+	if (ret)
+		return ret;
 
-	if (ct->opts.dst_addr) {
-		ret =
-		    pp_getaddrinfo(ct->opts.dst_addr, ct->ctrl_port, &results);
-		if (ret != 0)
-			return ret;
+	for (rp = results; rp; rp = rp->ai_next) {
+		ct->ctrl_connfd = socket(rp->ai_family, rp->ai_socktype,
+					 rp->ai_protocol);
+		if (ct->ctrl_connfd == -1) {
+			errno_save = -errno;
+			continue;
+		}
 
-		for (rp = results; rp != NULL; rp = rp->ai_next) {
-			ct->ctrl_connfd = socket(rp->ai_family, rp->ai_socktype,
-						 rp->ai_protocol);
-			if (ct->ctrl_connfd == -1)
-				continue;
+		if (ct->opts.src_port != 0) {
+			in_addr.sin_family = AF_INET;
+			in_addr.sin_port = htons(ct->opts.src_port);
+			in_addr.sin_addr.s_addr = htonl(INADDR_ANY);
 
-			memset(&ct->ctrl_addr, '\0', sizeof(ct->ctrl_addr));
-			ct->ctrl_addr.sin_family = AF_INET;
-			ct->ctrl_addr.sin_addr.s_addr =
-			    ((struct sockaddr_in *)rp->ai_addr)
-				->sin_addr.s_addr;
-			ct->ctrl_addr.sin_port = htons(ct->ctrl_port);
-
-			retp = inet_ntop(
-			    AF_INET,
-			    (const void *)&ct->ctrl_addr.sin_addr.s_addr, s,
-			    INET_ADDRSTRLEN);
-			if (retp == NULL) {
-				err = -errno;
-				PP_PRINTERR("inet_ntop", err);
-				return err;
-			}
-			PP_DEBUG("CLIENT: connecting to <%s> (%s)\n",
-				 ct->opts.dst_addr, s);
-			ret = connect(ct->ctrl_connfd,
-				      (struct sockaddr *)&ct->ctrl_addr,
-				      sizeof(ct->ctrl_addr));
+			ret =
+			    bind(ct->ctrl_connfd, (struct sockaddr *)&in_addr,
+				 sizeof(in_addr));
 			if (ret == -1) {
-				err = -errno;
+				errno_save = -errno;
 				close(ct->ctrl_connfd);
 				continue;
 			}
-			break;
 		}
-		if (ct->ctrl_connfd == -1) {
-			err = -errno;
-			PP_PRINTERR("getaddrinfo/socket/connect", err);
-			return err;
-		}
-		PP_DEBUG("CLIENT: connected\n");
+
+		ret = connect(ct->ctrl_connfd, rp->ai_addr, rp->ai_addrlen);
+		if (ret != -1)
+			break;
+
+		errno_save = -errno;
+		close(ct->ctrl_connfd);
+	}
+
+	if (!rp || ret == -1) {
+		ret = errno_save;
```

> If results == NULL, then errno_save will be undefined.

right...i guess i should check that first and fail early.